pandas/tests/dtypes/test_inference.py

```diff
@@ -531,6 +531,15 @@ def test_maybe_convert_objects_uint64(self):
         exp = np.array([2 ** 63, -1], dtype=object)
         tm.assert_numpy_array_equal(lib.maybe_convert_objects(arr), exp)
 
+    def test_maybe_convert_objects_datetime(self):
+        # GH27438
+        arr = np.array([0, 0], dtype=object)
+        arr[0] = np.datetime64('2000-01-01')
```

Comment:
this construction seems odd, you can just construct directly no? e.g.
```arr = np.array([np.datetime...., np.timedetla.lll], dtype='object'])```

Reply:
don't you need a NaT somewhere? what about throwing in a np.nan as well (maybe another case)

Reply:
Construction is odd, yes - will rewrite (it's leftover of previous tests sorry). Will add tests with NaT as well. 

As for np.nan - not quite understand. Add something with datetimes and np.nan?

Reply:
yes we likely already have a test for this, but pls check

Reply:
@jreback all is green please check.

Now there are 10 tests for `maybe_convert_objects` + 13 tests through `Dataframe` constructor (there are also tests for `maybe_convert_numeric` but I did not count these).

Reply:
Edit by the way - no one confirm my changes with array full of `NaT` here - now it returns `datetime` without any warning of ambiguity with `timedelta`. Check it pls too.